### PR TITLE
Fix driver stuck on ext4 raid1 volume on Debian 9

### LINKED ISSUE #146: Tests are stucking on RAID on Debian 9 with kernel 4.9
Debian 9, raid, kernel 4.9, fs ext2, test test_transition_active_snapshot
```
[  215.753176] elastio-snap: initializing tracer
[  215.753176] elastio-snap: creating kernel cow thread
[  215.753232] elastio-snap: getting the base block device's make_request_fn
[  215.753242] elastio-snap: freezing 'md6'
[  215.825139] md: md6: resync done.
[  215.838726] RAID1 conf printout:
[  215.838728]  --- wd:2 rd:2
[  215.838729]  disk 0, wo:0, o:1, dev:loop0p1
[  215.838731]  disk 1, wo:0, o:1, dev:loop1p1
[  215.840878] elastio-snap: starting tracing
[  215.840880] elastio-snap: thawing 'md6'
[  215.841327] elastio-snap: stopping cow thread
[  363.484118] INFO: task python3:26423 blocked for more than 120 seconds.
[  363.485146]       Tainted: G           O    4.9.0-18-amd64 #1 Debian 4.9.303-1
[  363.486117] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[  363.487144] python3         D    0 26423  26400 0x00000000
[  363.487149]  0000000000000086 ffff9b9def9e4000 0000000000000000 ffff9b9da8f30000
[  363.487152]  ffff9b9dffd5da00 ffff9b9df36a6040 ffffafc7c27efc00 ffffffffbaa1eeb9
[  363.487153]  000000000001e480 00ff9b9dffd5e480 ffff9b9dffd5da00 ffffffffbb04df60
[  363.487156] Call Trace:
[  363.487163]  [<ffffffffbaa1eeb9>] ? __schedule+0x239/0x6f0
[  363.487165]  [<ffffffffbaa1f3a2>] ? schedule+0x32/0x80
[  363.487167]  [<ffffffffbaa2284d>] ? schedule_timeout+0x1dd/0x380
[  363.487170]  [<ffffffffba4d3726>] ? vprintk_emit+0x316/0x4d0
[  363.487171]  [<ffffffffbaa1fde1>] ? wait_for_completion+0xf1/0x130
[  363.487173]  [<ffffffffba4a66a0>] ? wake_up_q+0x70/0x70
[  363.487175]  [<ffffffffba49b8a5>] ? kthread_stop+0x45/0xf0
[  363.487182]  [<ffffffffc0aa98c0>] ? __tracer_destroy_cow_thread.isra.44+0x20/0x40 [elastio_snap]
[  363.487184]  [<ffffffffc0aabe25>] ? ioctl_transition_inc+0x175/0x460 [elastio_snap]
[  363.487186]  [<ffffffffbaa137d2>] ? printk+0x5a/0x76
[  363.487189]  [<ffffffffba5842f5>] ? find_get_pages_tag+0x155/0x2d0
[  363.487191]  [<ffffffffc0aadce3>] ? ctrl_ioctl+0x693/0xc30 [elastio_snap]
[  363.487192]  [<ffffffffba61f1a5>] ? do_filp_open+0xa5/0x100
[  363.487194]  [<ffffffffba5bc97d>] ? handle_mm_fault+0x87d/0x1350
[  363.487196]  [<ffffffffba622492>] ? do_vfs_ioctl+0xa2/0x620
[  363.487197]  [<ffffffffba622a84>] ? SyS_ioctl+0x74/0x80
[  363.487199]  [<ffffffffba403b7d>] ? do_syscall_64+0x8d/0x100
[  363.487201]  [<ffffffffbaa23d8e>] ? entry_SYSCALL_64_after_swapgs+0x58/0xc6
[  363.487208] INFO: task elastio_snap_in:27010 blocked for more than 120 seconds.
[  363.488246]       Tainted: G           O    4.9.0-18-amd64 #1 Debian 4.9.303-1
[  363.489237] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[  363.490248] elastio_snap_in D    0 27010      2 0x00000000
[  363.490252]  0000000000000046 ffff9b9def9e4000 ffff9b9def9e4000 ffff9b9df15aa040
[  363.490255]  ffff9b9dffd5da00 ffff9b9da8f30000 ffffafc7c2eb7eb0 ffffffffbaa1eeb9
[  363.490257]  ffffffffba4bdd6f 00000000baa23ea5 ffff9b9dffd5da00 ffffafc7c27efc38
[  363.490259] Call Trace:
[  363.490265]  [<ffffffffbaa1eeb9>] ? __schedule+0x239/0x6f0
[  363.490267]  [<ffffffffba4bdd6f>] ? __wake_up_common+0x4f/0x90
[  363.490272]  [<ffffffffc0aaa4a0>] ? __cow_write_mapping+0x140/0x140 [elastio_snap]
[  363.490273]  [<ffffffffbaa1f3a2>] ? schedule+0x32/0x80
[  363.490275]  [<ffffffffba49b0ec>] ? kthread+0xac/0xf0
[  363.490277]  [<ffffffffbaa23eb1>] ? __switch_to_asm+0x41/0x70
[  363.490278]  [<ffffffffba49b040>] ? kthread_park+0x60/0x60
[  363.490280]  [<ffffffffbaa23f37>] ? ret_from_fork+0x57/0x70
[  484.316170] INFO: task python3:26423 blocked for more than 120 seconds.
[  484.323487]       Tainted: G           O    4.9.0-18-amd64 #1 Debian 4.9.303-1
[  484.325531] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[  484.327215] elastio_snap_in D    0 27010      2 0x00000000
[  484.327218]  0000000000000046 ffff9b9def9e4000 ffff9b9def9e4000 ffff9b9df15aa040
[  484.327220]  ffff9b9dffd5da00 ffff9b9da8f30000 ffffafc7c2eb7eb0 ffffffffbaa1eeb9
[  484.327221]  ffffffffba4bdd6f 00000000baa23ea5 ffff9b9dffd5da00 ffffafc7c27efc38
[  484.327223] Call Trace:
[  484.327229]  [<ffffffffbaa1eeb9>] ? __schedule+0x239/0x6f0
[  484.327231]  [<ffffffffba4bdd6f>] ? __wake_up_common+0x4f/0x90
[  484.327235]  [<ffffffffc0aaa4a0>] ? __cow_write_mapping+0x140/0x140 [elastio_snap]
[  484.327236]  [<ffffffffbaa1f3a2>] ? schedule+0x32/0x80
[  484.327239]  [<ffffffffba49b0ec>] ? kthread+0xac/0xf0
[  484.327240]  [<ffffffffbaa23eb1>] ? __switch_to_asm+0x41/0x70
[  484.327241]  [<ffffffffba49b040>] ? kthread_park+0x60/0x60
[  484.327242]  [<ffffffffbaa23f37>] ? ret_from_fork+0x57/0x70
```

The problem is very similar to the #143 and #145. And, probably, the fix will be the same. But let it be created as a separate bug because of any fs and raid and 4.9 kernel.

### PR DESCRIPTION
**Pull request by e-kov**
TL;DR
==
This is a reflection of fix 49826d1 recently made by @skypodolsky
for the LVM/RAID setups and driver stuck in the loop of the
`snap_handle_write_bio` function. This time the same fix is applied for
the read bios in the `snap_handle_read_bio` function.

Long description of the problem
==
An investigation was started from the hung of the `test_modify_origin`
unit-test on Debian 9 with Linux kernel 4.9 on volume `raid1` and `ext4`
file system.
The test contains the following steps:
1. setup snapshot of the raid volume
2. do some writes on the original volume
3. os.sync()
4. mount snapshot device
5. ...

The 4th operation led to a hang of the `mount` operation. Interesting
detail, the hang was not reproducible without `os.sync()` in the test or
on `xfs` fs (because `xfs` doesn't honour sync operation).
The `mount` operation performs reads from the device. And in our case we
had infinite iterations inside of the `bio_for_each_segment` -> `while`
loop in the `snap_handle_read_bio` function. And `snap_cow_thread` was
loading 1 core for 100%. Usage of the `bio_for_each_segment_all` allows
to iterate via all segments of the bio and fix this problem.
Another interesting point: potentially this bug may occur with different
kernel versions in cases where the read bio is split into multiple
segments. But the most favorable conditions for this turned out to be
precisely after the sync operation before reading (mount call) on
kernel 4.9 with raid1.

Resolves #146

### CODE REVIEW

**Comment by anelson** [comment]
@e-kov, GitHub-flavored Markdown tip for you:

When you write a commit message in Git, the `#` character gets interpreted as a comment delimiter so you can't make headers that way, as you've probably noticed:

```markdown
## Header <= This will be stripped from the commit message by `git` because it's a comment

Some text...
```

When I compose commit messages that will become PR descriptions, I use the alternative way of designating a header, which is to place the `=` character on the following line, one for every header level you want to create:

```markdown
Also A Header
==

Some text...
```

You can see how GitHub renders that:

> Also A Header
> ==
> 
> Some text...

You can do this instead of the `**=` you've used here.

**Review by anelson — APPROVED** [review]
Great, detailed PR description and great catch!

**Review by skypodolsky — APPROVED** [review]

**Review by vsazhenyuk-softheme — APPROVED** [review]
